rts/lra/coordinator/src/main/java/io/narayana/lra/coordinator/internal/LRARecoveryModule.java

```diff
@@ -200,53 +242,69 @@ public void getRecoveringLRAs(Map<URI, LongRunningAction> lras) {
      * @return false if record isn't in the store or there was an error removing it
      */
     public boolean removeCommitted(Uid lraUid) {
-        try {
-            return _recoveryStore.remove_committed(lraUid, _transactionType);
-        } catch (ObjectStoreException e) {
-            if (LRALogger.logger.isTraceEnabled()) {
-                LRALogger.logger.tracef(e,
-                        "LRARecoveryModule: Object store exception '%s' while removing LRA record %s",
-                        e.getMessage(), lraUid.fileStringForm());
-            } else if (LRALogger.logger.isInfoEnabled()) {
-                LRALogger.logger.infof(
-                        "LRARecoveryModule: Object store exception '%s' while removing LRA record %s",
-                        e.getMessage(), lraUid.fileStringForm());
+        synchronized (this) {
+            try {
+                return _recoveryStore.remove_committed(lraUid, _transactionType) ||
+                        _recoveryStore.remove_committed(lraUid, LongRunningAction.getFailedLraType());
+            } catch (ObjectStoreException e) {
+                if (LRALogger.logger.isTraceEnabled()) {
+                    LRALogger.logger.tracef(e,
+                            "LRARecoveryModule: Object store exception '%s' while removing LRA record %s",
+                            e.getMessage(), lraUid.fileStringForm());
+                } else if (LRALogger.logger.isInfoEnabled()) {
+                    LRALogger.logger.infof(
+                            "LRARecoveryModule: Object store exception '%s' while removing LRA record %s",
+                            e.getMessage(), lraUid.fileStringForm());
+                }
             }
-        }
 
-        return false;
+            return false;
+        }
     }
 
     public void getFailedLRAs(Map<URI, LongRunningAction> lras) {
         InputObjectState aa_uids = new InputObjectState();
-        Consumer<Uid> failedLRACreator = uid -> {
-            LongRunningAction lra = new LongRunningAction(lraService, new Uid(uid));
-            lra.activate();
+        String failedTransactionType = LongRunningAction.getFailedLraType();
 
-            LRAStatus status = lra.getLRAStatus();
-            if (LRAStatus.FailedToCancel.equals(status) || LRAStatus.FailedToClose.equals(status)) {
-                lras.put(lra.getId(), lra);
+        synchronized (this) {
```

Comment:
Is the `synchronized` necessary here? What's the danger of not synchronized access as it was used before?

Reply:
Here, since the records are now fetched from two different locations, but earlier we were fetching the records from single location. So, `synchronized` could be useful here in the race condition.

Reply:
We should not be adding `synchronized` at place where we think "it could be useful". When I'm looking at this I believe it could not be necessary as the recovery will eventually finalize the state to be consistent. The only danger which I can think about (please correct me) is a possible duplicity in the output?

Reply:
Yes, it could print duplicate records, if not marked `synchronized`

Reply:
I'm really not sure if it's good idea to add `synchronize` here. Why to slow down processing when the consistency issue that are considered here is only possible duplicity - which I can't observe either how that could be for the achieved (?!) with the used data structure of {{Map}}. Use a data structure that eliminates the duplicity and that's it, or am I wrong in my assumption?

Reply:
Correct, but there could also be a scenario in which thread calls get failed logs, and at the same time anoher thread deletes the failed logs, at that point of time, we are returning some logs to the user, but actually there is no record present in the system.

So if the mentioned point is not of much concern, I can remove the `synchronized` keyword.

Reply:
Yes, you are right. Those are good points you say. Let me scratch another two examples to continue the discussion.

Imagine that a thread gets the failed logs and it gets out of the synchronized block back to the REST endpoint but before the HTTP I/O response is sent the thread is suspended. The thread scheduler assigns the time to the recovery thread which deletes the failed logs. Does the `synchronized` helped here? Do you have some chance to defend against such "inconsistency"?

Another question. The remote API asks for failed LRAs, the coordinator returns it and the program saves that to internal map. It starts to iterate over it and starts printing it to the console. Meanwhile the coordinator recovery deletes the failed logs. Does the `synchronized` helped here? Do you have some chance to defend against such "inconsistency"?

Reply:
Okay, in your scenarios, `synchronized` keyword would not help.
I got your point, that the scenario of inconsistency could be possible by other means. So just `synchronizing` at the particular point would not help much, but instead make the runs slow.
I will remove this `synchronized` block then.